tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,382 @@
+            slowmo_optim.slowmo_factor * initial_slow_momentum_buffer
+            + (initial_prev_params - fsdp_net.params[0].data) / 0.01
+        )
+
+        # parameter_(t+1) = prev_param - slowmo_lr * base_lr * momentum_(t+1)
+        calculated_params = initial_prev_params - 0.1 * 0.01 * momentum
+
+        self.assertEqual(fsdp_net_slowmo.params[0].data, calculated_params)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("sharding_strategy", [ShardingStrategy.NO_SHARD])
+    def test_slowmo_optimizer_state_dict(
+        self, sharding_strategy: Optional[ShardingStrategy]
+    ):
+        chkpt = tempfile.gettempdir() + "/checkpoint.pt"
+        fsdp_net_slowmo = FSDP(
+            Net(has_wrapping=False, sharding_strategy=sharding_strategy),
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        ).to(self.rank)
+
+        cur_subgroup = dist.new_group(ranks=[self.rank])
+        slowMoState = slow_momentum_comm.SlowMoState(cur_subgroup)
+        fsdp_net_slowmo.register_comm_hook(slowMoState, slow_momentum_comm.slowmo_hook)
+        inpt = torch.randn(7, 8).float()
+        inpt.to(self.rank)  # type: ignore[call-overload]
+
+        slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+            base_optim=torch.optim.SGD(fsdp_net_slowmo.parameters(), lr=1e-2),
+            slowmo_freq=4,
+            slowmo_factor=0.5,
+            slowmo_lr=0.1,
+        )
+
+        for _ in range(10):
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim)
+
+        state = {"optim_state_dict": slowmo_optim.state_dict()}
+
+        if self.rank == 0:
+            torch.save(state, chkpt)
+
+        dist.barrier()
+
+        map_location = {"cuda:%d" % 0: "cuda:%d" % self.rank}
+        checkpoint = torch.load(chkpt, map_location=map_location)
+
+        slowmo_optim_dummy = slow_momentum_optimizer.SlowMomentumOptimizer(
+            base_optim=torch.optim.SGD(fsdp_net_slowmo.parameters(), lr=1e-2),
+            slowmo_freq=2,
+            slowmo_factor=3,
+            slowmo_lr=4,
+        )
+        slowmo_optim_dummy.load_state_dict(checkpoint["optim_state_dict"])
+        # make sure acerager's period and step were updated
+        self.assertEqual(
+            slowmo_optim_dummy.averager.period, slowmo_optim.averager.period
+        )
+        self.assertEqual(slowmo_optim_dummy.averager.step, slowmo_optim.averager.step)
+
+        # make sure slowmo parameters were updated
+        self.assertEqual(slowmo_optim_dummy.slowmo_freq, slowmo_optim.slowmo_freq)
+        self.assertEqual(slowmo_optim_dummy.slowmo_factor, slowmo_optim.slowmo_factor)
+        self.assertEqual(slowmo_optim_dummy.slowmo_lr, slowmo_optim.slowmo_lr)
+
+        for _ in range(10):
+            self._train_step(inpt, fsdp_net_slowmo, slowmo_optim_dummy)
+
+        self.assertEqual(slowmo_optim_dummy.averager.step, 20)
+
+    @skip_if_lt_x_gpu(2)
+    def test_slowmo_optimizer_warnings(self):
```

Comment:
nit: maybe more aptly named as `test_slowmo_optimizer_errors` since everything is an error?